// @lobehub/ui/es/ThemeProvider/ThemeProvider.mjs
'use client';

import { useCdnFn } from "../ConfigProvider/index.mjs";
import { generateCustomStylish } from "../styles/theme/customStylish.mjs";
import { generateCustomToken } from "../styles/theme/customToken.mjs";
import FontLoader_default from "../FontLoader/index.mjs";
import { LOBE_THEME_APP_ID } from "./constants.mjs";
import { createLobeAntdTheme } from "../styles/theme/antdTheme.mjs";
import ConfigProvider_default from "./ConfigProvider.mjs";
import GlobalStyle_default from "./GlobalStyle/index.mjs";
import { memo, useCallback, useLayoutEffect, useMemo, useRef } from "react";
import { Fragment as Fragment$1, jsx, jsxs } from "react/jsx-runtime";
import { App } from "antd";
import { ThemeProvider } from "antd-style";
import { merge } from "es-toolkit/compat";

//#region src/ThemeProvider/ThemeProvider.tsx
const CSS_VAR_PREFIX = "css-var-";
const ThemeProvider$1 = memo(({ children, customStylish, customToken, enableCustomFonts = true, enableGlobalStyle = true, customFonts, customTheme = {}, className, style, theme: antdTheme, ...rest }) => {
	const genCdnUrl = useCdnFn();
	const appRef = useRef(null);
	useLayoutEffect(() => {
		const node = appRef.current;
		if (!node) return;
		const htmlEl = document.documentElement;
		let currentClasses = [];
		const syncCssVarClasses = () => {
			for (const cls of currentClasses) htmlEl.classList.remove(cls);
			const newClasses = [];
			let el$1 = node;
			while (el$1 && el$1 !== htmlEl) {
				for (const cls of el$1.classList) if (cls.startsWith(CSS_VAR_PREFIX)) newClasses.push(cls);
				el$1 = el$1.parentElement;
			}
			for (const cls of newClasses) htmlEl.classList.add(cls);
			currentClasses = newClasses;
		};
		syncCssVarClasses();
		const observer = new MutationObserver(syncCssVarClasses);
		let el = node;
		while (el && el !== htmlEl) {
			observer.observe(el, { attributeFilter: ["class"] });
			el = el.parentElement;
		}
		return () => {
			observer.disconnect();
			for (const cls of currentClasses) htmlEl.classList.remove(cls);
		};
	}, []);
	const webfontUrls = useMemo(() => customFonts || [
		genCdnUrl({
			path: "css/index.css",
			pkg: "@lobehub/webfont-mono"
		}),
		genCdnUrl({
			path: "css/index.css",
			pkg: "@lobehub/webfont-harmony-sans"
		}),
		genCdnUrl({
			path: "css/index.css",
			pkg: "@lobehub/webfont-harmony-sans-sc"
		}),
		genCdnUrl({
			path: "dist/katex.min.css",
			pkg: "katex"
		})
	], [customFonts, genCdnUrl]);
	const stylish = useCallback((theme$1) => ({
		...generateCustomStylish(theme$1),
		...customStylish?.(theme$1)
	}), [customStylish]);
	const token = useCallback((theme$1) => ({
		...generateCustomToken(theme$1),
		...customToken?.(theme$1)
	}), [customToken]);
	const theme = useCallback((appearance) => {
		return merge(createLobeAntdTheme({
			appearance,
			neutralColor: customTheme.neutralColor,
			primaryColor: customTheme.primaryColor
		}), antdTheme);
	}, [
		customTheme.primaryColor,
		customTheme.neutralColor,
		antdTheme
	]);
	return /* @__PURE__ */ jsxs(Fragment$1, { children: [enableCustomFonts && webfontUrls?.length > 0 && webfontUrls.map((webfont) => /* @__PURE__ */ jsx(FontLoader_default, { url: webfont }, webfont)), /* @__PURE__ */ jsx(ThemeProvider, {
		customStylish: stylish,
		customToken: token,
		theme,
		...rest,
		children: /* @__PURE__ */ jsxs(ConfigProvider_default, { children: [enableGlobalStyle && /* @__PURE__ */ jsx(GlobalStyle_default, {}), /* @__PURE__ */ jsx(App, {
			className,
			style: {
				isolation: "isolate",
				minHeight: "inherit",
				width: "inherit",
				...style
			},
			children: /* @__PURE__ */ jsx("div", {
				id: LOBE_THEME_APP_ID,
				ref: appRef,
				style: { display: "contents" },
				children
			})
		})] })
	})] });
});
ThemeProvider$1.displayName = "LobeThemeProvider";
var ThemeProvider_default = ThemeProvider$1;

//#endregion
export { ThemeProvider_default as default };
